R/runner.R

```diff
@@ -226,11 +256,71 @@ orderly_runner_ <- R6::R6Class(
     #' Submit an arbitrary job on the queue
     #'
     #' @param job A quoted R expression.
-    #' @param environment Environment to run the expression in.
+    #' @param depends_on Task ids for any dependencies of this job.
     #'
     #' @return Task id
-    submit = function(job, environment = parent.frame()) {
-       self$queue$enqueue_(job, environment)
+    submit = function(expr, depends_on = NULL) {
+      self$queue$enqueue_(expr, depends_on = depends_on)
+    },
+
+    #' @description
+    #' Queue a workflow.
+    #'
+    #' @param reports Details of reports to be run.
+    #' @param ref The git sha to run the workflow.
+    #' @param changelog Description of changes to the reports - applied to
+    #' all reports.
+    #' @param poll How frequently to poll for the report ID being available.
+    #' Applied to each of the reports run in the workflow
+    #' @param timeout Timeout for each of the reports run as part
+    #' of the workflow default 3 hours.
+    #'
+    #' @return The key for the workflow and each individual report
+    submit_workflow = function(reports, ref = NULL, changelog = NULL,
+                               poll = 0.1, timeout = 60 * 60 * 3) {
+      self$assert_ref_switching_allowed(ref)
+      if (!is.null(ref)) {
+        dependency_root <- self$alternative_root
+        git_fetch(dependency_root)
+        prev <- git_checkout_branch(ref, root = dependency_root)
+        on.exit(git_checkout_branch(prev, root = dependency_root))
+      } else {
+        dependency_root <- self$root
+      }
+      workflow <- build_workflow(self$root, dependency_root, reports)
+      ## Build a list of dependencies
+      ## report_name : [task_id1, task_id2, ...]
+      ## There could be multiple tasks queued with the same name
+      dependencies <- key_value_collector()
+      queue_task <- function(report) {
+        depends_on <- dependencies$get(report$depends_on)
+        key <- self$submit_task_report(name = report$name,
+                                           parameters = report$params,
+                                           ref = ref,
+                                           instance = report$instance$source,
```

Comment:
If I'm reading this correctly then instance is expected to be an object ([schema](https://github.com/vimc/orderly.server/pull/66/files#diff-b04094791cbba4533fa79ccb4687d0b7727a0b48106a1f3b7d66c80ee06fb22bL8)) but the only value that's used is the one with key `source`. This is fine - and I guess somewhat futureproof - we'll be sure to only send requests of this form:
```
{"name": "<report_name>", "instance": {"source": "<instance_name>"}}
```
Just noting here for comparison with `endpoint_run` where we send a query param `?instance=<instance_name>`

Reply:
Yeah so happy to go either way with this. Can see argument to use a query param like we do for running a single report for consistency. There is a ticket to move instance into the json body https://mrc-ide.myjetbrains.com/youtrack/issue/VIMC-4598 for running so maybe ok to accept the inconsistency for now. Any thoughts @richfitz

Reply:
Sorry misunderstood that, yeah maybe going to a string for now is better? Removes that knowledge from orderly.server about `support` being a special. And having this work with multiple instances is probably not going to happen any time soon.

Reply:
I'm don't have a strong (or very well-informed!) opinion on this. I'm happy to leave as-is (i.e. we hard-code `source` in our app, just allowing users to specify instance of that database), which I guess also gives you the flexibility to support multiple databases without subsequently changing the orderly.server API. But will follow whatever you choose.

Reply:
BTW if it stays as an object then I think it would be preferable for schema to use `instances` rather than `instance`

Reply:
Thanks, have made this change to take a string. I think better to remove that hardcoding of `source` here and in OW. Can make this an object proper when we need to support multiple instances.